src/mlpack/methods/gmm/diagonal_gmm.hpp

```diff
@@ -0,0 +1,308 @@
+ *     const arma::vec& probabilities,
+ *     std::vector<distribution::DiagonalGaussianDistribution>& dists,
+ *     arma::vec& weights);
+ * @endcode
+ *
+ * Example use:
+ *
+ * @code
+ * // Set up a mixture of 5 gaussians in a 4-dimensional space.
+ * DiagonalGMM g(5, 4);
+ *
+ * // Train the DiagonalGMM given the data observations, using the default
+ * // EM fitting mechanism.
+ * 
+ * g.Train(data);
+ *
+ * // Get the probability of 'observation' being observed from this
+ * // DiagoanlGMM.
+ * double probability = g.Probability(observation);
+ *
+ * // Get a random observation from the DiagonalGMM.
+ * arma::vec observation = g.Random();
+ * @endcode
+ */
+class DiagonalGMM
+{
+ private:
+  //! The number of Gaussians in the model.
+  size_t gaussians;
+  //! The dimensionality of the model.
+  size_t dimensionality;
+
+  //! Vector of Gaussians
+  std::vector<distribution::DiagonalGaussianDistribution> dists;
+
+  //! Vector of a priori weights for each Gaussian.
+  arma::vec weights;
+
+ public:
+  /**
+   * Create an empty Diagonal Gaussian Mixture Model, with zero gaussians.
+   */
+  DiagonalGMM() :
+      gaussians(0),
+      dimensionality(0)
+  {
+    // Warn the user.  They probably don't want to do this.  If this
+    // constructor is being used (because it is required by some template
+    // classes), the user should know that it is potentially dangerous.
+    Log::Debug << "DiagonalGMM::DiagonalGMM(): no parameters given;"
+        "Estimate() may fail " << "unless parameters are set." << std::endl;
+  }
+
+  /**
+   * Create a GMM with the given number of Gaussians, each of which have the
+   * specified dimensionality.  The means and covariances will be set to 0.
+   *
+   * @param gaussians Number of Gaussians in this DiagonalGMM.
+   * @param dimensionality Dimensionality of each Gaussian.
+   */
+  DiagonalGMM(const size_t gaussians, const size_t dimensionality);
+
+  /**
+   * Create a DiagonalGMM with the given dists and weights.
+   *
+   * @param dists Distributions of the model.
+   * @param weights Weights of the model.
+   */
+  DiagonalGMM(const std::vector<distribution::DiagonalGaussianDistribution>&
+      dists, const arma::vec& weights) :
+      gaussians(dists.size()),
+      dimensionality((!dists.empty()) ? dists[0].Mean().n_elem : 0),
+      dists(dists),
+      weights(weights) { /* Nothing to do. */ }
+
+  //! Copy constructor for DiagonalGMMs.
+  DiagonalGMM(const DiagonalGMM& other);
+
+  //! Copy operator for DiagonalGMMs.
+  DiagonalGMM& operator=(const DiagonalGMM& other);
+
+  //! Return the number of gaussians in the model.
+  size_t Gaussians() const { return gaussians; }
+  //! Return the dimensionality of the model.
+  size_t Dimensionality() const { return dimensionality; }
+
+  /**
+   * Return a const reference to a component distribution.
+   *
+   * @param i index of component.
+   */
+  const distribution::DiagonalGaussianDistribution& Component(size_t i) const {
```

Comment:
Can you put the opening ``{``on a new line, just to be consistent with the rest of the codebase.